test/metric_value_benchmark/collectd_test.go

```diff
@@ -72,41 +59,58 @@ func (t *CollectDTestRunner) validateCollectDMetric(metricName string) status.Te
 		Status: status.FAILED,
 	}
 
-	dims, failed := t.DimensionFactory.GetDimensions([]dimension.Instruction{
+	instructions := []dimension.Instruction{
 		{
 			Key:   "InstanceId",
 			Value: dimension.UnknownDimensionValue(),
 		},
-		{
-			Key:   "type_instance",
-			Value: dimension.ExpectedDimensionValue{aws.String("user")},
-		},
-		{
-			Key:   "instance",
-			Value: dimension.ExpectedDimensionValue{aws.String("0")},
-		},
-		{
+	}
+	switch metricName {
+	case "collectd_counter_1_value":
+		instructions = append(instructions, dimension.Instruction{
+			// CWA adds this metric_type dimension.
 			Key:   "type",
-			Value: dimension.ExpectedDimensionValue{aws.String("percent")},
-		},
-	})
+			Value: dimension.ExpectedDimensionValue{Value: aws.String("counter")},
+		})
+	case "collectd_gauge_1_value":
+		instructions = append(instructions, dimension.Instruction{
+			// CWA adds this metric_type dimension.
+			Key:   "type",
+			Value: dimension.ExpectedDimensionValue{Value: aws.String("gauge")},
+		})
+	}
 
+	dims, failed := t.DimensionFactory.GetDimensions(instructions)
 	if len(failed) > 0 {
 		return testResult
 	}
-
 	fetcher := metric.MetricValueFetcher{}
+	// Namespace must match the JSON config.
 	values, err := fetcher.Fetch(namespace, metricName, dims, metric.AVERAGE, test_runner.HighResolutionStatPeriod)
+
 	if err != nil {
 		return testResult
 	}
 
-	if !isAllValuesGreaterThanOrEqualToZero(metricName, values) {
+	runDuration := t.GetAgentRunDuration()
+	// aggregationInterval must match the JSON configuration.
+	aggregationInterval := 30 * time.Second
+	// If aggregation is not happening there could be a data point every 5 seconds.
+	// So validate the upper bound.
+	upperBound := int(runDuration/aggregationInterval) + 2
+	// Allow 2 missing data points in case CW-Metrics-Web-Service has a 1 minute
+	// delay to store.
```

Comment:
? This is not the SLA for metric put-to-get. We should not need this amount of variance.

Reply:
In high resources consumption environment, I would disagree with you; however, if we only running to validate then I agree this should be acceptable

Reply:
No. You're talking about how long it takes to get a response from CloudWatch. What is the resource constrained environment?